PowerEditor/src/Parameters.cpp

```diff
@@ -1939,9 +1943,417 @@ bool NppParameters::getUserStylersFromXmlTree()
 {
 	TiXmlNode *root = _pXmlUserStylerDoc->FirstChild(L"NotepadPlus");
 	if (!root) return false;
+	updateUserStylersFromModelXml(root);
 	return feedStylerArray(root);
 }
 
+void NppParameters::updateKeyWordsFromModelXml(TiXmlNode* rootUser)
+{
+	std::wstring modelLangsPath(_nppPath);
+	pathAppend(modelLangsPath, L"langs.model.xml");
+	TiXmlDocument* pXmlModel = new TiXmlDocument(modelLangsPath);
+	int buffer_size = WideCharToMultiByte(CP_ACP, 0, modelLangsPath.c_str(), static_cast<int>(modelLangsPath.length()), nullptr, 0, nullptr, nullptr);
+	std::string sModelPath(buffer_size, 0);
+	WideCharToMultiByte(CP_ACP, 0, modelLangsPath.c_str(), static_cast<int>(modelLangsPath.length()), sModelPath.data(), buffer_size, nullptr, nullptr);
+
+	// if there's a problem loading the model XML, just exit out (don't need to warn the user, since the main Langs XML has already been loaded
+	//	the same logic will be used for any other errors while trying to do this XML merge
+	if (!pXmlModel->LoadFile())
+		return;
+
+	TiXmlNode* rootModel = pXmlModel->FirstChild(L"NotepadPlus");
+	if (!rootModel)
+		return;
+
+	// now that the model is reasonable, it's reasonable to do the MD5-checking
+	MD5 md5;
+	std::string md5digest_model = md5.digestFile(sModelPath.c_str());
+	buffer_size = MultiByteToWideChar(CP_UTF8, 0, md5digest_model.c_str(), static_cast<int>(md5digest_model.length()), nullptr, 0);
+	std::wstring wsDigest(buffer_size, 0);
+	MultiByteToWideChar(CP_UTF8, 0, md5digest_model.c_str(), static_cast<int>(md5digest_model.length()), wsDigest.data(), buffer_size);
+
+	std::string sUserText{};
+	_pXmlDoc->Print(sUserText);
+	std::string md5digest_user_text_before = md5.digestString(sUserText.c_str());
+
+	// if modelMD5 is the same as the one seen in the XML, don't need to merge in the model...
+	TiXmlElement* peRootUser = rootUser->ToElement();
+	const wchar_t* pwct_modelMD5 = peRootUser->Attribute(L"modelMD5");
+	buffer_size = WideCharToMultiByte(CP_UTF8, 0, pwct_modelMD5 ? pwct_modelMD5 : L"", -1, nullptr, 0, nullptr, nullptr);
+	std::string s_modelMD5_from_xml(buffer_size, 0);
+	WideCharToMultiByte(CP_UTF8, 0, pwct_modelMD5 ? pwct_modelMD5 : L"", -1, s_modelMD5_from_xml.data(), buffer_size, nullptr, nullptr);
+	s_modelMD5_from_xml.pop_back();	// remove the NULL-terminator
+	if (md5digest_model == s_modelMD5_from_xml)
+		return;
+
+	// update (or add) the MD5 stored in the XML
+	peRootUser->SetAttribute(L"modelMD5", wsDigest.c_str());
+
+	// get the <Languages> element from both user and model langs
+	TiXmlElement* langsTopUser = rootUser->FirstChildElement(L"Languages");
+	TiXmlElement* langsTopModel = rootModel->FirstChildElement(L"Languages");
+	if (!langsTopUser || !langsTopModel)
+		return;
+
+	// map each of the user-file's languages -> element-pointer, to keep track of the languages already in the user-file
+	map<std::wstring, TiXmlElement*> mapUserLanguages{};

```

Comment:
for consistency I would probably use qualified `std::map` since it is used for other containers (string, vector, and even for map elsewhere), Similar to other WinAPI usage in Notepad++, global scope for `::WideCharToMultiByte` and `::MultiByteToWideChar`